### ISSUE TITLE
Oauth2: cannot fetch token: 401 Unauthorized

### ISSUE THREAD

**Issue by eelbrink** [hashicorp/terraform-provider-googleworkspace#228]
Hi there,

I have am experiencing an issue which i am hoping you can assist me with. I am looking forward to using this provider for google workspace user and group admin. 

### Terraform Version

> Terraform v1.0.11
> on darwin_arm64
> provider registry.terraform.io/hashicorp/googleworkspace v0.5.2

### Affected Resource(s)
Please list the resources as a list, for example:

> google workspace user and group administration

### Terraform Configuration Files
```hcl
terraform {
  required_providers {
    googleworkspace = {
      source = "hashicorp/googleworkspace"
      version = "0.5.2"
    }
  }
}

provider "googleworkspace" {
  # Configuration options
  customer_id             = "xxxxxxxxxx"
  credentials             = "/Users/xxxxxxx/Desktop/terraform-key.json"
  impersonated_user_email = "xxxxxxxx@xxxxxxxx"
}

```

### Actual Behavior
```
➜  GoogleWorkspace terraform apply        
╷
│ Error: Get "https://admin.googleapis.com/admin/directory/v1/users/xxxxxxx%40xxxxxxxxx?alt=json&prettyPrint=false": oauth2: cannot fetch token: 401 Unauthorized
│ Response: {
│   "error": "unauthorized_client",
│   "error_description": "Client is unauthorized to retrieve access tokens using this method, or client not authorized for any of the scopes requested."
│ }
│ 
│   with data.googleworkspace_user.xxxx,
│   on check_users.tf line 1, in data "googleworkspace_user" "xxxxx":
│    1: data "googleworkspace_user" "xxxxx" {
```

### Documentation Followed

- https://www.hashicorp.com/blog/announcing-the-google-workspace-provider-for-hashicorp-terraform-tech-preview#groupssettings-googleapis-com
- https://developers.google.com/admin-sdk/directory/v1/guides/delegation#create_the_service_account_and_credentials
- https://developers.google.com/admin-sdk/directory/v1/guides/delegation#delegate_domain-wide_authority_to_your_service_account

**Comment by megan07** [comment]
Hi @eelbrink! I'm so happy you're looking forward to using this, and I'm sorry this confusion is your first experience. I'm currently trouble-shooting a similar issue (maybe the same?) [here](https://github.com/hashicorp/terraform-provider-googleworkspace/issues/213#issuecomment-992957190). I would suggest reviewing the scopes you added for your service account when setting up Domain Wide Delegation. If you aren't specifying scopes in the provider, it defaults to all of them, which, currently is:
```	"https://www.googleapis.com/auth/gmail.settings.basic",
	"https://www.googleapis.com/auth/gmail.settings.sharing",
	"https://www.googleapis.com/auth/chrome.management.policy",
	"https://www.googleapis.com/auth/cloud-platform",
	"https://www.googleapis.com/auth/admin.directory.customer",
	"https://www.googleapis.com/auth/admin.directory.domain",
	"https://www.googleapis.com/auth/admin.directory.group",
	"https://www.googleapis.com/auth/admin.directory.orgunit",
	"https://www.googleapis.com/auth/admin.directory.rolemanagement",
	"https://www.googleapis.com/auth/admin.directory.userschema",
	"https://www.googleapis.com/auth/admin.directory.user",
	"https://www.googleapis.com/auth/apps.groups.settings",
```

Does this help? If not, I'll continue to try and reproduce the issue.

**Comment by megan07** [comment]
I should add, if you _are_ specifying scopes in the provider, then make sure that the scopes added in Domain Wide Delegation include all of the scopes you're adding to the provider.

**Comment by megan07** [comment]
@eelbrink - I wanted to reach out and see if this is still an issue and if there's anything more I can help with. Thanks!

**Comment by paoll** [comment]
Hello,

I'm joining that issue as I have the exact same issue with the following use case: retrieving users belonging to set of groups, and extract their customSchema.

As stated in the documentation, `impersonate` is mandatory as Directory API is not opened to service Account. 
So I do have a SA GCP side, with a domain wide delegation enabled, and almost all scopes possible authorized. I'm using a super admin account to impersonate, thus, I have the same error message:

```
Response: {
    "error": "unauthorized_client",
    "error_description": "Client is unauthorized to retrieve access tokens using this method, or client not authorized for any of the scopes requested."
}
```

The odd thing is that exact same SA and service account are used in another program without any issue. 

Here is my terraform configuration:

```
provider "googleworkspace" {
	credentials = var.credentials_path
        impersonated_user_email = "super-admin@company.com"
       customer_id             = "XXXXXXX"
}

data "googleworkspace_user" "google_groups_user" {

	primary_email = "randomuser@company.com"
}
```

And versions used:
```
provider: 0.6.0
terraform: 1.1.6
```

Many thanks in case you have any lead !

**Comment by paoll** [comment]
> I've also spent hours trying to get authentication working.
> 
> Finally I have a working solution now using a service account and `oauth_scopes`:
> 
> The important piece is that the scopes for the service account and the `oauth_scopes` in the TF provider module must match.
> 
> I.e. if you set
> 
> ```
> provider "googleworkspace" {
>   credentials = [...]
>   customer_id = [...]
>   impersonated_user_email = [...]
>   oauth_scopes = [
>     "https://www.googleapis.com/auth/admin.directory.group",
>     "https://www.googleapis.com/auth/admin.directory.user"
>   ]
> ```
> 
> Then **exactly** these scopes needs to be defined in https://admin.google.com/ac/owl/domainwidedelegation for the respective service account.

Thank you so much @pat-s , I had a subset of scopes declared on TF side, but more scopes declared on Google side.
Works like a charm now !

**Comment by jshcmpbll** [comment]
This should be in the provider docs..

**Comment by SarahFrench** [comment]
Hi everyone- this issue appears to be resolved so I'm going to close it. @eelbrink if you still need any help please open a new issue!

I thought I'd post a summary of the underlying problem for anyone looking at this issue in future.

The issues described above appear to revolve around [this note in the documentation](https://registry.terraform.io/providers/hashicorp/googleworkspace/latest/docs#:~:text=Note%3A%20The%20Oauth%20scopes%20granted%20to%20your%20service%20account%20must%20match%2C%20or%20be%20a%20superset%2C%20of%20the%20oauth_scopes%20granted%20to%20the%20googleworkspace%20provider.). Here's the note:

>Note: The Oauth scopes granted to your service account must match, or be a superset, of the oauth_scopes granted to the googleworkspace provider.

# Summary

Here I've written out some examples of configurations that will work and won't work due to oath scopes, and why that is.
Expand the section below to see:

<details>
  <summary>❗ Click to expand ❗</summary>
  
## Example of something that works

If you're using domain-wide delegation you need to make sure the `oauth_scopes` used to configure the provider are a subset of scopes given to the service account on the [domain-wide delegation page](https://admin.google.com/u/1/ac/owl/domainwidedelegation).

E.g. If I have a service account with these 2 scopes:
- https://www.googleapis.com/auth/admin.directory.user
- https://www.googleapis.com/auth/admin.directory.userschema
 
<img width="300" alt="A service account set up as a client for domain-wide delegation with 2 scopes: admin.directory.user, and admin.directory.userschema" src="https://user-images.githubusercontent.com/15078782/170327714-2e2a237c-74c8-4a43-a71e-00a41b8e9547.png">

Then I can access data about users. In fact only the `https://www.googleapis.com/auth/admin.directory.user` scope is needed for what I want to achieve in the Terraform configuration below, so I'll configure the provider with only that scope:

```
# This configuration has the provider authorised to access a pre-existing Workspace
# that contains users not controlled by this configuration

terraform {
  required_providers {
    googleworkspace = {
      source  = "hashicorp/googleworkspace"
      version = "0.6.0"
    }
  }
}

provider "googleworkspace" {
  credentials             = "/relative-path/from-root-module/credentials.json"
  customer_id             = "xxxxxxx"
  impersonated_user_email = "xxxxxxx@xxxxxxx.com"
  oauth_scopes = [
      "https://www.googleapis.com/auth/admin.directory.user"
  ]
}

data "googleworkspace_users" "user-list" {}

output "num_users" {
  value = length(data.googleworkspace_users.user-list.users)
}

```

> INFO: the above configuration works with `oauth_scopes` set to just that 1 scope, and also works if it contains the both of the scopes the service account has (but the 2nd scope is not necessary for the configuration to function) 

## Example of something slightly different that doesn't work

If I edit the configuration above so the provider is configured with an additional scope, e.g. scope for groups:

```diff
provider "googleworkspace" {
  credentials             = "/foo/bar/relative-path/credentials.json"
  customer_id             = "xxxxxxx"
  impersonated_user_email = "xxxxxxx@xxxxxxx.com"
  oauth_scopes = [
      "https://www.googleapis.com/auth/admin.directory.user"
+      "https://www.googleapis.com/auth/admin.directory.group"

  ]
}
```

This will result in the `oauth2: cannot fetch token: 401 Unauthorized` error experienced by the people in this issue **because the provider tries to be issued a token that included an additional scope that the service account isn't authorised to use**. You could fix this by removing the extra unnecessary scope from the Terraform configuration, or adding that scope to the service account on the [domain-wide delegation page](https://admin.google.com/u/1/ac/owl/domainwidedelegation).

## The problem with not including `oauth_scopes` in the provider configuration

If you don't include `oauth_scopes` in the provider configuration then it [defaults to requesting these 12 scopes](https://github.com/hashicorp/terraform-provider-googleworkspace/blob/e900a62b4d97b90a22bfb2bddd2dce293f9c0f00/internal/provider/provider.go#L17-L30). This can result in a situation like above, where the provider is attempting to use more scopes than the service account is authorised to use.

**For your configuration to work when `oauth_scopes` is not explicitly set in your configuration, your service account will need to have at least these 12 scopes enabled.**

```diff
provider "googleworkspace" {
  credentials             = "/foo/bar/relative-path/credentials.json"
  customer_id             = "xxxxxxx"
  impersonated_user_email = "xxxxxxx@xxxxxxx.com"
}

# is equivalent to

provider "googleworkspace" {
  credentials             = "/foo/bar/relative-path/credentials.json"
  customer_id             = "xxxxxxx"
  impersonated_user_email = "xxxxxxx@xxxxxxx.com"
  oauth_scopes = [
    "https://www.googleapis.com/auth/gmail.settings.basic",
    "https://www.googleapis.com/auth/gmail.settings.sharing",
    "https://www.googleapis.com/auth/chrome.management.policy",
    "https://www.googleapis.com/auth/cloud-platform",
    "https://www.googleapis.com/auth/admin.directory.customer",
    "https://www.googleapis.com/auth/admin.directory.domain",
    "https://www.googleapis.com/auth/admin.directory.group",
    "https://www.googleapis.com/auth/admin.directory.orgunit",
    "https://www.googleapis.com/auth/admin.directory.rolemanagement",
    "https://www.googleapis.com/auth/admin.directory.userschema",
    "https://www.googleapis.com/auth/admin.directory.user",
    "https://www.googleapis.com/auth/apps.groups.settings",
  ]
}

```

</details>

# TLDR

If you don't include `oauth_scopes` in your provider configuration then it [defaults to requesting these 12 scopes](https://github.com/hashicorp/terraform-provider-googleworkspace/blob/e900a62b4d97b90a22bfb2bddd2dce293f9c0f00/internal/provider/provider.go#L17-L30). Unless your service account is set up with at least those 12 scopes you will experience the `oauth2: cannot fetch token: 401 Unauthorized` error described above, because the provider is trying to use more scopes than the service account has authorisation to use.

If you include `oauth_scopes` in your provider configuration it will resolve the problem by making your configuration more explicit. Just make sure that the scopes you specify there are included in the scopes the service account has authorisation to use :)

**Comment by jshcmpbll** [comment]
Thanks for looking into it further! I think the instructions around setting up access to those 12 scopes could be defined better in the documentation. I thought I had properly setup access for all the necessary scopes and still received this error which was quite frustrating. This was all for a throw away workspace and we were just validating the maturity of the provider so we were basically throwing every scope we could at it and could not get past this error..

- https://registry.terraform.io/providers/hashicorp/googleworkspace/latest/docs#authorization

My assumption is maybe one of the scopes is missing from the domain wide delegation but its been a while since we tried getting this to work. I'll revisit it soon and post my findings in a new issue/PR if I find anything different.

**Comment by SarahFrench** [comment]
Hi @jshcmpbll, I agree that the docs could use some more details about this and I hope to get an update out soon! And please do let us know if you have issues, or discover anything you think could be documented more clearly.
